sympy/calculus/util.py

```diff
@@ -1649,3 +1552,129 @@ def union(self, other):
 
 # setting an alias for AccumulationBounds
 AccumBounds = AccumulationBounds
+
+def get_preprocessor(cls):
```

Comment:
I don't understand the purpose of this function that encloses the two others that are the actual functions that we want...

Reply:
Frankly speaking, `AccumBounds` does not need preprocessor since its behavior is already hard-coded in `Add.flatten`. I pushed the last three commits to show that with preprocessor, any class can use `Add` and `Mul` as constructor without modifying`flatten`.  

Again, the reason why I am proposing the preprocessor is to let user define his/her own class and make it compatible with `Add` or `Mul`, without meddling their `__new__`, `flatten` or `_from_args`.

Reply:
I referred to matrix module's `get_postprocessor` function and used its design. Should I discard this wrapper function? I can easily do it.

Reply:
I believe the matrices one is written like this because the two processors were so similar that it made sense to combine them into a single function. It looks like that isn't the case here, as the Add and Mul parts don't share any code.

Reply:
Fixed it.